lib/winapi/winnt.c

```diff
@@ -0,0 +1,26 @@
+#include <winnt.h>
+#include <xboxkrnl/xboxkrnl.h>
+
+LONG64 InterlockedExchange64 (LONG64 volatile *Target, LONG64 Value)
+{
+    // There's no xchg8b in x86, so we implement it with cmpxchg8b
+    LARGE_INTEGER li;
+    li.QuadPart = *Target;
+    __asm__ __volatile__("1:\n"
+                         "cmpxchg8b (%2)\n"
+                         "jnz 1b\n" // If the comparison failed, try again with updated edx:eax
```

Comment:
The comment make it look like "there can go something wrong". But from my understanding, it has to fail the first round by design. I would like to see a different/extended documentation of the overall asm code. It took me way to long to figure out whats going on here. (And now: Hoping that I didn't make myself a fool for this 😅)

Reply:
Nah it does usually work first try. But it may fail if the thread gets interrupted between the read from `Target`, from which edx and eax get initialized when entering the asm block, and `cmpxchg8b`, and the other thread modifies the target.

Reply:
To wrap my head around it:
- eax:edx will contain the Target-value(as in *not* the address)
- ebx:ecx will contain the Value that should be placed in the Target soon
- r contains the address of Target

Your code does:
1. load up the register as stated above
2. `cmpxchg8b` compares the value at the address of r(the value of Target) with the content of eax:eax
3. If equal: copy ebx:ecx into the memory pointed to by r(setting new value of Target)
4. If not equal: jump back to step 2.

So I would agree now, that it won't fail most of the time. But, I don't get the point of this function in that case. 😟

Reply:
> But, I don't get the point of this function in that case.

Wdym? It's a winapi function to atomically exchange two 8-byte values.

Reply:
[after clarifying via discord]
Instead of my original question, to further document the asm instruction, I would propose to add your thoughts to your first comment in Line 6. 
- Want to swap bytes atomic
- No instruction available 
- could do it via dis- & reenable interrupts
- opted for "abusing" `cmpxchg8b` due to guessing its faster(and simpler)

Anyway, I think thats up to you. Keep in mind that I can't resolve this, as it is just a comment.